File: apps/relay-server/src/routes/stream.py
"""WebSocket 스트리밍 엔드포인트.

App ↔ Relay Server: /relay/calls/{call_id}/stream
  - App에서 User 오디오/텍스트를 수신
  - App으로 자막/번역 오디오/상태 알림을 전송
"""

import json
import logging

from fastapi import APIRouter, WebSocket, WebSocketDisconnect

from src.call_manager import call_manager
from src.types import WsMessage, WsMessageType

router = APIRouter(tags=["stream"])
logger = logging.getLogger(__name__)


@router.websocket("/calls/{call_id}/stream")
async def app_websocket(ws: WebSocket, call_id: str):
    """App ↔ Relay Server WebSocket 연결.

    User의 오디오/텍스트를 받아 Session A로 전달하고,
    Session B의 번역 결과를 App으로 전달한다.
    """
    await ws.accept()
    logger.info("App WebSocket connected (call=%s)", call_id)

    call = call_manager.get_call(call_id)
    if not call:
        await ws.send_json(
            WsMessage(
                type=WsMessageType.ERROR,
                data={"message": "Call not found"},
            ).model_dump()
        )
        await ws.close()
        return

    # App WS를 call_manager에 등록 (AudioRouter가 이 WS로 메시지 전송)
    call_manager.register_app_ws(call_id, ws)

    # AudioRouter가 아직 없으면 Twilio 연결 대기 중
    if not call_manager.get_router(call_id):
        try:
            await ws.send_json(
                WsMessage(
                    type=WsMessageType.CALL_STATUS,
                    data={"status": "waiting", "message": "전화 연결 중..."},
                ).model_dump()
            )
        except Exception:
            pass

    try:
        while True:
            raw = await ws.receive_text()

            try:
                msg = json.loads(raw)
            except json.JSONDecodeError:
                logger.warning("Invalid JSON from App WS (call=%s)", call_id)
                continue

            msg_type = msg.get("type", "")

            audio_router = call_manager.get_router(call_id)
            if not audio_router:
                continue

            match msg_type:
                case "audio_chunk":
                    audio_b64 = msg.get("data", {}).get("audio", "")
                    if audio_b64:
                        await audio_router.handle_user_audio(audio_b64)
                case "vad_state":
                    state = msg.get("data", {}).get("state", "")
                    if state == "committed":
                        await audio_router.handle_user_audio_commit()
                case "text_input":
                    text = msg.get("data", {}).get("text", "")
                    if text:
                        await audio_router.handle_user_text(text)
                case "typing_state":
                    await audio_router.handle_typing_started()
                case "end_call":
                    logger.info("User ended call via WebSocket (call=%s)", call_id)
                    break

    except WebSocketDisconnect:
        logger.info("App WebSocket disconnected (call=%s)", call_id)
    except Exception as e:
        logger.error("App WebSocket error (call=%s): %s", call_id, e)
    finally:
        await call_manager.cleanup_call(call_id, reason="app_disconnected")

File: apps/relay-server/src/routes/stream.ts
import type { FastifyInstance } from 'fastify';
import type WebSocket from 'ws';
import { activeCalls, endActiveCall } from './calls.js';
import { AudioRouter } from '../realtime/audio-router.js';
import { FirstMessageHandler } from '../realtime/first-message.js';
import { checkGuardrail, createGuardrailEvent, getFillerPhrase } from '../guardrail/checker.js';
import { config } from '../config.js';
import type { ClientMessage, ServerMessage } from '../types.js';

export async function streamRoute(app: FastifyInstance) {
  /**
   * WS /relay/calls/:id/stream — Realtime audio/text streaming
   */
  app.get<{ Params: { id: string } }>(
    '/calls/:id/stream',
    { websocket: true },
    (socket, request) => {
      const callId = request.params.id;
      const existing = activeCalls.get(callId);

      if (!existing) {
        sendToClient(socket, {
          type: 'error',
          code: 'CALL_NOT_FOUND',
          message: 'Active call not found',
        });
        socket.close();
        return;
      }

      const { call, sessionManager, ringBuffer, costTracker, guardrailEvents } = existing;
      let audioRouter: AudioRouter | null = null;
      let firstMessageHandler: FirstMessageHandler | null = null;
      let warningTimer: ReturnType<typeof setTimeout> | null = null;
      let idleTimer: ReturnType<typeof setTimeout> | null = null;

      app.log.info(`[Stream] Client connected for call ${callId}`);

      // Notify client of current call status
      sendToClient(socket, { type: 'call.status', status: call.status });

      // ── Setup Session Manager Callbacks ──
      // (Reconnect callbacks to route to this WebSocket client)

      const callbacks = {
        onTranscriptUser: (text: string, language: typeof call.sourceLanguage) => {
          sendToClient(socket, { type: 'transcript.user', text, language, timestamp: Date.now() });
          resetIdleTimer();
        },
        onTranscriptUserTranslated: async (text: string, language: typeof call.targetLanguage) => {
          // Run guardrail check on user's translated text (going to recipient)
          const result = await checkGuardrail(text, language);
          guardrailEvents.push(createGuardrailEvent(result));

          if (result.level === 3 && !result.passed) {
            // Level 3 failed: block transmission, notify user
            sendToClient(socket, {
              type: 'guardrail.triggered',
              level: 3,
              original: text,
            });
            return;
          }

          const finalText = result.correctedText ?? text;
          sendToClient(socket, { type: 'transcript.user.translated', text: finalText, language, timestamp: Date.now() });

          if (result.level >= 2 && result.correctedText) {
            sendToClient(socket, {
              type: 'guardrail.triggered',
              level: result.level,
              original: text,
              corrected: result.correctedText,
            });
          }
        },
        onTranscriptRecipient: (text: string, language: typeof call.targetLanguage) => {
          sendToClient(socket, { type: 'transcript.recipient', text, language, timestamp: Date.now() });
          resetIdleTimer();
        },
        onTranscriptRecipientTranslated: async (text: string, language: typeof call.sourceLanguage) => {
          // Run guardrail on recipient's translated text (going to user)
          const result = await checkGuardrail(text, language);
          guardrailEvents.push(createGuardrailEvent(result));

          const finalText = result.correctedText ?? text;
          sendToClient(socket, { type: 'transcript.recipient.translated', text: finalText, language, timestamp: Date.now() });
        },
        onAudioForTwilio: (audioBase64: string) => {
          ringBuffer.push(audioBase64, 'user');
          audioRouter?.routeSessionAToTwilio(audioBase64);
        },
        onAudioForApp: (audioBase64: string) => {
          sendToClient(socket, { type: 'audio.recipient.translated', audio: audioBase64 });
        },
        onSessionError: (session: 'A' | 'B', error: string) => {
          sendToClient(socket, { type: 'error', code: 'SESSION_ERROR', message: `Session ${session}: ${error}` });
          // Trigger recovery
          sendToClient(socket, { type: 'session.recovery', status: 'recovering', gapMs: 0 });
        },
        onRecipientSpeechStarted: () => {
          audioRouter?.handleRecipientSpeechStart();
          sendToClient(socket, { type: 'interrupt.detected', source: 'recipient' });
          // First message trigger: recipient answered the phone
          firstMessageHandler?.handleRecipientDetected();
        },
        onRecipientSpeechEnded: () => {
          audioRouter?.handleRecipientSpeechEnd();
        },
        onFirstMessageComplete: () => {
          firstMessageHandler?.handleFirstMessageComplete();
        },
      };

      // Patch callbacks into session manager
      // (SessionManager was created with placeholder callbacks in calls.ts,
      //  now we connect the real callbacks via the internal handlers)
      Object.assign((sessionManager as unknown as { callbacks: typeof callbacks }).callbacks, callbacks);

      // ── Setup Call Duration Warning ──
      const elapsed = Date.now() - call.startedAt;
      const warningIn = config.callWarningAtMs - elapsed;
      if (warningIn > 0) {
        warningTimer = setTimeout(() => {
          const remaining = config.maxCallDurationMs - (Date.now() - call.startedAt);
          sendToClient(socket, {
            type: 'call.warning',
            remainingMs: remaining,
            message: `통화 종료까지 ${Math.ceil(remaining / 60_000)}분 남았습니다`,
          });
        }, warningIn);
      }

      // ── Idle Timer ──
      function resetIdleTimer() {
        if (idleTimer) clearTimeout(idleTimer);
        call.lastActivityAt = Date.now();
        idleTimer = setTimeout(() => {
          sendToClient(socket, {
            type: 'call.warning',
            remainingMs: 0,
            message: '30초간 대화가 없었습니다. 통화를 종료할까요?',
          });
        }, config.callIdleTimeoutMs);
      }

      resetIdleTimer();

      // ── Handle Client Messages ──
      socket.on('message', (data: unknown) => {
        try {
          const message: ClientMessage = JSON.parse(
            typeof data === 'string' ? data : (data as Buffer).toString(),
          );

          switch (message.type) {
            case 'audio.chunk':
              // User audio → Session A (voice-to-voice mode)
              ringBuffer.push(message.audio, 'user');
              audioRouter?.routeUserAudioToSessionA(message.audio);
              resetIdleTimer();
              break;

            case 'audio.commit':
              // Client VAD: end of speech
              audioRouter?.commitUserAudio();
              break;

            case 'text.send':
              // Push-to-Talk: text → Session A → TTS → Twilio
              audioRouter?.routeUserTextToSessionA(message.text);
              resetIdleTimer();
              break;

            case 'vad.speech_start':
              // Client VAD notification
              break;

            case 'vad.speech_end':
              // Client VAD notification
              break;

            case 'call.end':
              // User requested call end
              endActiveCall(callId, 'completed');
              sendToClient(socket, { type: 'call.status', status: 'completed' });
              cleanup();
              break;
          }
        } catch (err) {
          app.log.error({ err }, '[Stream] Failed to parse client message');
        }
      });

      socket.on('close', () => {
        app.log.info(`[Stream] Client disconnected for call ${callId}`);
        cleanup();
      });

      socket.on('error', (err: Error) => {
        app.log.error({ err }, `[Stream] WebSocket error for call ${callId}`);
      });

      function cleanup() {
        if (warningTimer) clearTimeout(warningTimer);
        if (idleTimer) clearTimeout(idleTimer);
        firstMessageHandler?.cleanup();
        audioRouter = null;
        firstMessageHandler = null;
      }
    },
  );
}

function sendToClient(ws: WebSocket, message: ServerMessage) {
  if (ws.readyState === ws.OPEN) {
    ws.send(JSON.stringify(message));
  }
}
